File: fraiseql-swift/Sources/FraiseQL/Schema.swift
import Foundation

typealias FieldConfig = [String: Any]
typealias TypeFields = [String: FieldConfig]

/// Validator for field-level scope format and patterns
///
/// Scope format: action:resource
/// Examples: read:user.email, admin:*, write:Post.*
///
/// Rules:
/// - Action: [a-zA-Z_][a-zA-Z0-9_]*
/// - Resource: [a-zA-Z_][a-zA-Z0-9_.]*|*
enum ScopeValidator {
  /// Validates scope format: action:resource
  ///
  /// - Parameter scope: The scope string to validate
  /// - Returns: true if valid, false otherwise
  static func validate(_ scope: String) -> Bool {
    if scope.isEmpty {
      return false
    }

    if scope == "*" {
      return true
    }

    let parts = scope.split(separator: ":", maxSplits: 1)
    guard parts.count == 2 else {
      return false
    }

    let action = String(parts[0])
    let resource = String(parts[1])

    guard !action.isEmpty, !resource.isEmpty else {
      return false
    }

    return isValidAction(action) && isValidResource(resource)
  }

  /// Validates a list of scopes
  ///
  /// - Parameter scopes: The list of scopes to validate
  /// - Returns: true if all are valid, false otherwise
  static func validateAll(_ scopes: [String]) -> Bool {
    guard !scopes.isEmpty else { return false }
    return scopes.allSatisfy(validate)
  }

  /// Checks if action matches pattern: [a-zA-Z_][a-zA-Z0-9_]*
  private static func isValidAction(_ action: String) -> Bool {
    guard !action.isEmpty else { return false }
    let first = action.first!
    guard first.isLetter || first == "_" else { return false }
    return action.dropFirst().allSatisfy { $0.isLetter || $0.isNumber || $0 == "_" }
  }

  /// Checks if resource matches pattern: [a-zA-Z_][a-zA-Z0-9_.]*|*
  private static func isValidResource(_ resource: String) -> Bool {
    if resource == "*" { return true }
    guard !resource.isEmpty else { return false }
    let first = resource.first!
    guard first.isLetter || first == "_" else { return false }
    return resource.dropFirst().allSatisfy { $0.isLetter || $0.isNumber || $0 == "_" || $0 == "." }
  }
}

struct FieldDefinition: Codable {
  let name: String
  let type: String
  let nullable: Bool
}

struct TypeInfo {
  let name: String
  let fields: TypeFields
  let description: String?
}

final class SchemaRegistry {
  static let shared = SchemaRegistry()
  private var types: [String: TypeInfo] = [:]
  private let lock = NSLock()

  private init() {}

  func register(_ name: String, _ info: TypeInfo) {
    lock.lock()
    defer { lock.unlock() }
    types[name] = info
  }

  func getTypeNames() -> [String] {
    lock.lock()
    defer { lock.unlock() }
    return Array(types.keys)
  }

  func getType(_ name: String) -> TypeInfo? {
    lock.lock()
    defer { lock.unlock() }
    return types[name]
  }

  func clear() {
    lock.lock()
    defer { lock.unlock() }
    types.removeAll()
  }
}

public enum Schema {
  public static func registerType(_ name: String, fields: TypeFields, description: String? = nil) {
    let info = TypeInfo(name: name, fields: fields, description: description)
    SchemaRegistry.shared.register(name, info)
  }

  public static func exportTypes(pretty: Bool = true) -> String {
    let registry = SchemaRegistry.shared
    let typeNames = registry.getTypeNames()

    var types: [[String: Any]] = []
    for typeName in typeNames {
      guard let typeInfo = registry.getType(typeName) else { continue }

      var fieldsArray: [[String: Any]] = []
      for (fieldName, fieldConfig) in typeInfo.fields {
        let fieldType = fieldConfig["type"] as? String ?? "String"
        let nullable = fieldConfig["nullable"] as? Bool ?? false

        var field: [String: Any] = [
          "name": fieldName,
          "type": fieldType,
          "nullable": nullable
        ]

        // Add scope if present
        if let scope = fieldConfig["requiresScope"] as? String {
          field["requiresScope"] = scope
        }

        // Add scopes if present
        if let scopes = fieldConfig["requiresScopes"] as? [String] {
          field["requiresScopes"] = scopes
        }

        fieldsArray.append(field)
      }

      var typeObj: [String: Any] = [
        "name": typeName,
        "fields": fieldsArray
      ]
      if let desc = typeInfo.description {
        typeObj["description"] = desc
      }
      types.append(typeObj)
    }

    let schema: [String: Any] = ["types": types]
    let options: JSONSerialization.WritingOptions = pretty ? [.prettyPrinted, .sortedKeys] : []

    if let jsonData = try? JSONSerialization.data(withJSONObject: schema, options: options),
       let jsonString = String(data: jsonData, encoding: .utf8) {
      return jsonString
    }
    return "{\"types\":[]}"
  }

  public static func exportTypesFile(_ path: String) {
    let json = exportTypes(pretty: true)
    do {
      let fileURL = URL(fileURLWithPath: path)
      try fileURL.parent().createDirectory(withIntermediateDirectories: true, attributes: nil)
      try json.write(toFile: path, atomically: true, encoding: .utf8)

      let count = SchemaRegistry.shared.getTypeNames().count
      print("✅ Types exported to \(path)")
      print("   Types: \(count)")
      print()
      print("🎯 Next steps:")
      print("   1. fraiseql compile fraiseql.toml --types \(path)")
      print("   2. This merges types with TOML configuration")
      print("   3. Result: schema.compiled.json with types + all config")
    } catch {
      fatalError("Failed to write types file: \(path)")
    }
  }

  public static func reset() {
    SchemaRegistry.shared.clear()
  }

  public static func getTypeNames() -> [String] {
    return SchemaRegistry.shared.getTypeNames()
  }

  public static func getType(_ name: String) -> TypeInfo? {
    return SchemaRegistry.shared.getType(name)
  }
}

extension URL {
  func parent() -> URL {
    return deletingLastPathComponent()
  }
}

File: fraiseql-dart/lib/schema.dart
import 'dart:convert';
import 'dart:io';

typedef FieldConfig = Map<String, dynamic>;
typedef TypeFields = Map<String, FieldConfig>;

class TypeInfo {
  final String name;
  final TypeFields fields;
  final String? description;

  TypeInfo({
    required this.name,
    required this.fields,
    this.description,
  });
}

class SchemaRegistry {
  static final SchemaRegistry _instance = SchemaRegistry._internal();
  final Map<String, TypeInfo> _types = {};

  factory SchemaRegistry() {
    return _instance;
  }

  SchemaRegistry._internal();

  void register(String name, TypeInfo info) {
    _types[name] = info;
  }

  List<String> getTypeNames() => _types.keys.toList();

  TypeInfo? getType(String name) => _types[name];

  void clear() => _types.clear();
}

/// Validates scope format: action:resource
/// Examples: read:user.email, admin:*, write:Post.*
bool _validateScope(String scope) {
  if (scope.isEmpty) return false;
  if (scope == '*') return true;

  final parts = scope.split(':');
  if (parts.length != 2) return false;

  final action = parts[0];
  final resource = parts[1];

  if (action.isEmpty || resource.isEmpty) return false;

  return _isValidAction(action) && _isValidResource(resource);
}

/// Validates action matches pattern: [a-zA-Z_][a-zA-Z0-9_]*
bool _isValidAction(String action) {
  if (action.isEmpty) return false;
  final first = action[0];
  if (!RegExp(r'[a-zA-Z_]').hasMatch(first)) return false;

  for (int i = 1; i < action.length; i++) {
    if (!RegExp(r'[a-zA-Z0-9_]').hasMatch(action[i])) return false;
  }
  return true;
}

/// Validates resource matches pattern: [a-zA-Z_][a-zA-Z0-9_.]*|*
bool _isValidResource(String resource) {
  if (resource == '*') return true;
  if (resource.isEmpty) return false;

  final first = resource[0];
  if (!RegExp(r'[a-zA-Z_]').hasMatch(first)) return false;

  for (int i = 1; i < resource.length; i++) {
    if (!RegExp(r'[a-zA-Z0-9_.]').hasMatch(resource[i])) return false;
  }
  return true;
}

class Schema {
  static final _registry = SchemaRegistry();

  static void registerType(
    String name,
    TypeFields fields, {
    String? description,
  }) {
    // Validate scope fields in all fields
    for (final fieldEntry in fields.entries) {
      final fieldConfig = fieldEntry.value;
      final hasScope = fieldConfig.containsKey('requires_scope');
      final hasScopes = fieldConfig.containsKey('requires_scopes');

      // Check for conflicting scope and scopes
      if (hasScope && hasScopes) {
        throw FormatException(
          'Field "${fieldEntry.key}" cannot have both requires_scope and requires_scopes',
        );
      }

      // Validate requires_scope if present
      if (hasScope) {
        final scope = fieldConfig['requires_scope'];
        if (scope is! String) {
          throw FormatException(
            'Field "${fieldEntry.key}" requires_scope must be a string',
          );
        }
        if (!_validateScope(scope)) {
          throw FormatException(
            'Field "${fieldEntry.key}" has invalid scope format: "$scope"',
          );
        }
      }

      // Validate requires_scopes if present
      if (hasScopes) {
        final scopes = fieldConfig['requires_scopes'];
        if (scopes is! List) {
          throw FormatException(
            'Field "${fieldEntry.key}" requires_scopes must be a list',
          );
        }
        if (scopes.isEmpty) {
          throw FormatException(
            'Field "${fieldEntry.key}" requires_scopes cannot be empty',
          );
        }
        for (final scope in scopes) {
          if (scope is! String) {
            throw FormatException(
              'Field "${fieldEntry.key}" requires_scopes contains non-string value',
            );
          }
          if (!_validateScope(scope)) {
            throw FormatException(
              'Field "${fieldEntry.key}" has invalid scope in requires_scopes: "$scope"',
            );
          }
        }
      }
    }

    _registry.register(
      name,
      TypeInfo(name: name, fields: fields, description: description),
    );
  }

  static String exportTypes({bool pretty = true}) {
    final typeNames = _registry.getTypeNames();
    final types = <Map<String, dynamic>>[];

    for (final typeName in typeNames) {
      final typeInfo = _registry.getType(typeName);
      if (typeInfo == null) continue;

      final fieldsArray = <Map<String, dynamic>>[];
      typeInfo.fields.forEach((fieldName, fieldConfig) {
        final field = <String, dynamic>{
          'name': fieldName,
          'type': fieldConfig['type'] ?? 'String',
          'nullable': fieldConfig['nullable'] ?? false,
        };

        // Add requires_scope if present
        if (fieldConfig.containsKey('requires_scope')) {
          field['requires_scope'] = fieldConfig['requires_scope'];
        }

        // Add requires_scopes if present
        if (fieldConfig.containsKey('requires_scopes')) {
          field['requires_scopes'] = fieldConfig['requires_scopes'];
        }

        fieldsArray.add(field);
      });

      final typeObj = {
        'name': typeName,
        'fields': fieldsArray,
      };
      if (typeInfo.description != null) {
        typeObj['description'] = typeInfo.description;
      }
      types.add(typeObj);
    }

    final schema = {'types': types};
    final json = jsonEncode(schema);

    if (pretty) {
      return JsonEncoder.withIndent('  ').convert(schema);
    }
    return json;
  }

  static void exportTypesFile(String outputPath) {
    try {
      final json = exportTypes(pretty: true);
      final file = File(outputPath);

      file.parent.createSync(recursive: true);
      file.writeAsStringSync(json);

      final typesCount = _registry.getTypeNames().length;
      print('✅ Types exported to $outputPath');
      print('   Types: $typesCount');
      print('');
      print('🎯 Next steps:');
      print('   1. fraiseql compile fraiseql.toml --types $outputPath');
      print('   2. This merges types with TOML configuration');
      print('   3. Result: schema.compiled.json with types + all config');
    } catch (e) {
      throw Exception('Failed to write types file: $outputPath');
    }
  }

  static void reset() => _registry.clear();

  static List<String> getTypeNames() => _registry.getTypeNames();

  static TypeInfo? getType(String name) => _registry.getType(name);
}
